components/support/migration/src/main/java/mozilla/components/support/migration/FennecSettingsMigrator.kt

```diff
@@ -0,0 +1,100 @@
+    private lateinit var fennecAppPrefs: SharedPreferences
+
+    @VisibleForTesting const val FENNEC_APP_SHARED_PREFS_NAME = "GeckoApp"
+    @VisibleForTesting const val FENIX_SHARED_PREFS_NAME = "fenix_preferences"
+    @VisibleForTesting const val FENNEC_PREFS_FHR_KEY = "android.not_a_preference.healthreport.uploadEnabled"
+    @VisibleForTesting const val FENNEC_PREFS_TELEMETRY_KEY = "toolkit.telemetry.enabled"
+    @VisibleForTesting const val FENIX_PREFS_TELEMETRY_KEY = "pref_key_telemetry"
+
+    /**
+     * Migrate all Fennec - Fenix common SharedPreferences.
+     */
+    fun migrateSharedPrefs(context: Context): Result<SettingsMigrationResult> {
+        fennecAppPrefs = context.getSharedPreferences(FENNEC_APP_SHARED_PREFS_NAME, MODE_PRIVATE)
+        fenixPrefs = context.getSharedPreferences(FENIX_SHARED_PREFS_NAME, MODE_PRIVATE)
+
+        if (fennecAppPrefs.all.isEmpty()) {
+            return Result.Failure(
+                SettingsMigrationException(
+                    SettingsMigrationResult.Failure.EmptyPreference(FENNEC_APP_SHARED_PREFS_NAME)
+                )
+            )
+        }
+
+        return Result.Success(migrateAppPreferences())
+    }
+
+    private fun migrateAppPreferences(): SettingsMigrationResult {
+        return migrateTelemetryOptInStatus(fennecAppPrefs, fenixPrefs)
+    }
+}
+
+/**
+ * Wraps [SettingsMigrationResult] in an exception so that it can be returned via [Result.Failure].
+ *
+ * @property failure Wrapped [SettingsMigrationResult] indicating exact failure reason.
+ */
+class SettingsMigrationException(val failure: SettingsMigrationResult.Failure) :
+    Exception(failure.toString())
+
+/**
+ * Result of Fennec settings migration.
+ */
+sealed class SettingsMigrationResult {
+    /**
+     * Successful setting migration.
+     */
+    object Success : SettingsMigrationResult()
+
+    /**
+     * Failures of the Fennec settings migration.
+     */
+    sealed class Failure : SettingsMigrationResult() {
+        /**
+         * Wanted SharedPreference file is not accessible.
+         *
+         * A particular SharedPreference file contains many different settings.
+         * If one such desired file is empty the final result of the settings migration
+         * cannot be guaranteed so the entire process should be considered as failed.
+         */
+        data class EmptyPreference(val preferenceName: String) : Failure() {
+            override fun toString(): String {
+                return "Empty preference: $preferenceName"
+            }
+        }
+    }
+}
+
+private fun migrateTelemetryOptInStatus(
+    fennecPrefs: SharedPreferences,
+    fenixPrefs: SharedPreferences
+): SettingsMigrationResult {
+    val isTelemetryEnabled = fennecPrefs.getBoolean(FENNEC_PREFS_TELEMETRY_KEY, false)
+    val isHealthReportEnabled = fennecPrefs.getBoolean(FENNEC_PREFS_FHR_KEY, false)
+    val isTelemetryEnabledOnFennec = isHealthReportEnabled && isTelemetryEnabled
```

Comment:
My understanding was that we will simply use the FHR value. That's documented in https://github.com/mozilla-mobile/fenix/issues/5525 - `We are going to map those two values to a single value when migrating - with the FHR setting taking precedence.`

Reply:
Indeed, thanks!